sdks/java/extensions/join-library/src/main/java/org/apache/beam/sdk/extensions/joinlibrary/Join.java

```diff
@@ -350,6 +368,276 @@ public void processElement(ProcessContext c) {
+  }
+
+  private static class RightUnionTagFn<K, V1, V2>
+      extends SimpleFunction<KV<K, V2>, KV<K, KV<V1, V2>>> {
+    @Override
+    public KV<K, KV<V1, V2>> apply(KV<K, V2> element) {
+      return KV.of(element.getKey(), KV.of(null, element.getValue()));
+    }
+  }
+
+  private static class TemporalInnerJoinFn<K, V1, V2>
+      extends DoFn<KV<K, KV<V1, V2>>, KV<K, KV<V1, V2>>> {
+
+    @StateId("left")
+    private final StateSpec<OrderedListState<V1>> leftStateSpec;
+
+    @StateId("right")
+    private final StateSpec<OrderedListState<V2>> rightStateSpec;
+
+    @TimerId("eviction")
+    private final TimerSpec evictionSpec = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    private final Duration temporalBound;
+    private final Duration evictionFrequency;
+    private final SimpleFunction<KV<V1, V2>, Boolean> compareFn;
+    private transient boolean evictionTimerInitialized;
+
+    @Setup
+    public void setup() {
+      evictionTimerInitialized = false;
+    }
+
+    protected TemporalInnerJoinFn(
+        final Coder<V1> leftCoder,
+        final Coder<V2> rightCoder,
+        final Duration temporalBound,
+        SimpleFunction<KV<V1, V2>, Boolean> compareFn) {
+      this.leftStateSpec = StateSpecs.orderedList(leftCoder);
+      this.rightStateSpec = StateSpecs.orderedList(rightCoder);
+      this.temporalBound = temporalBound;
+      this.compareFn = compareFn;
+      this.evictionFrequency =
+          temporalBound.getMillis() <= 4 ? Duration.millis(1) : temporalBound.dividedBy(4);
+    }
+
+    /**
+     * Finds an element in {@code search} state that satisfies {@code func} within the non-inclusive
+     * time range (timestamp - temporalBound, timestamp + temporalBound).
+     *
+     * @param timestamp Instant that scopes the match within temporalBound (i.e. midpoint).
+     * @param search State to search.
+     * @param func Join predicate.
+     * @param <T> Type of the state collection and return value.
+     * @return Matching element or null if none is found.
+     */
+    @Nullable
+    private <T> TimestampedValue<T> findMatch(
+        Instant timestamp,
+        OrderedListState<T> search,
+        Function<TimestampedValue<T>, Boolean> func) {
+      Iterable<TimestampedValue<T>> searchIterable =
+          search.readRange(timestamp.minus(temporalBound), timestamp.plus(temporalBound));
+      if (searchIterable != null) {
+        for (TimestampedValue<T> current : searchIterable) {
+          if (new Duration(current.getTimestamp(), timestamp).abs().isShorterThan(temporalBound)
+              && func.apply(current)) {
+            return current;
+          }
+        }
+      }
+      return null;
+    }
+
+    /**
+     * Removes an element from state. This method is the semantic equivalent of an
+     * OrderedListState#remove which does not exist.
+     */
+    private <T> void remove(OrderedListState<T> state, TimestampedValue<T> element) {
+      Instant upperBound = element.getTimestamp().plus(1);
```

Comment:
If you have multiple elements with the same timestamp, this will remove all of those elements. Is that desired?

Reply:
A few lines down the returned collection is iterated, filtering out the one for removal, adding the rest back to state.

Reply:
I'm going to change the semantics of the join, so that it doesn't produce a 1:1 match, which means having a remove is no longer necessary. It will take me a few days to make the change so no need to review for now. I'll update when the PR is ready.

Reply:
Done. Closing this.